include/circt/Dialect/Seq/SeqOps.td

```diff
@@ -115,3 +115,47 @@ def FirRegOp : SeqOp<"firreg",
     bool hasReset() { return !!getReset(); }
   }];
 }
+
+def HLMemOp : SeqOp<"hlmem", [
+     Symbol,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]> ]> {
+
+  let summary = "Instantiate a high-level memory.";
+  let description = "See the Seq dialect rationale for a longer description";
+
+  let arguments = (ins I1:$clk, SymbolNameAttr:$sym_name);
+  let results = (outs HLMemType:$handle);
+
+  let extraClassDeclaration = [{
+    HLMemType getMemType() { return getHandle().getType().cast<HLMemType>(); }
+  }];
+
+  let assemblyFormat = "$sym_name $clk attr-dict `:` type($handle)";
+}
+
+def ReadPortOp : SeqOp<"read", 
+  [DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+  let summary = "Structural read access to a seq.hlmem";
+  let arguments = (ins 
+    HLMemType:$memory,
+    Variadic<AnyType>:$addresses,
+    I1:$rdEn,
+    I64Attr:$latency
+  );
+  let results = (outs AnyType:$readData);
+  let hasCustomAssemblyFormat = 1;
```

Comment:
Could this be assembly format with `custom` segments? May or may not make sense.

Reply:
I don't think there's any parts of this which are actually complex to parse (warranting `custom`). The main issue with using the ODS format here is the variadic addresses which have type constraints on each value wrt. the width of the corresponding dimension of the memory. This is ''neatly'' handled by `memref` in that addresses are of `index` type (nothing to check).